### PR TITLE
[WIP] Playing around with grid persistence.

### PR DESCRIPTION
**Pull request by csarofeen**
Made some changes to the grid persistent branch trying to get perf starting with an arbitrary size I would hope would be straight forward to get perf on (not really certain it is). I hardcoded this size under the benchmark: `NvFuserScheduler_ResNext_BatchNorm_nhwc_fp16`.

### CODE REVIEW

**Comment by csarofeen** [comment]
I'm getting 142GB/s on this size right now, which is underwhelming, but is better than where it started at.

File: benchmarks/cpp/nvfuser/batch_norm_channels_last.cpp [hunk 1]
```diff
@@ -298,16 +298,16 @@ NVFUSER_BENCHMARK_DEFINE(
...
+    // ->Args({128, 64, 112})
+    // ->Args({128, 128, 56})
+    // ->Args({128, 256, 56})
+    // ->Args({128, 128, 56})
+    // ->Args({128, 256, 28})
+    // ->Args({128, 512, 28})
+    // ->Args({128, 512, 14})
+    // ->Args({128, 1024, 14})
+    // ->Args({128, 1024, 7})
+    ->Args({8, 4096, 64})
```
Comment (csarofeen) [inline benchmarks/cpp/nvfuser/batch_norm_channels_last.cpp:310]: TODO: Remove before considering to merge.

File: torch/csrc/jit/codegen/cuda/codegen.cpp [hunk 2]
```diff
@@ -1963,6 +1963,11 @@ class CudaKernelGenerator : private OptOutConstDispatch {
     const auto reduction_name =
         genFusedReductionName(alloc_fused_reduction->out()->view());
 
+    if (!alloced_reduction_classes.emplace(alloc_fused_reduction->out()->view())
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/codegen.cpp:2116]: With welford, in an unswitch section there was multiple definitions being created because the grid reduce was being used in both the then/else section of the code. We should go fix the actual insertion of the allocations, instead of WAR-ing the codegen.

File: torch/csrc/jit/codegen/cuda/fusion_segmenter.h [hunk 3]
```diff
@@ -411,7 +411,8 @@ class CombineReductions;
 
 //! Options to configure/debug candidate finder
 struct TORCH_CUDA_CU_API SegmentCandidateFinderOptions {
-  bool run_translate_welford = true;
+  // Temporarily disable by default for now. May want to revisit.
+  bool run_translate_welford = false;
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/fusion_segmenter.h:415]: We want to use welford for grid persistence, forcing translation off everywhere for now.

File: torch/csrc/jit/codegen/cuda/scheduler/normalization.cpp [hunk 4]
```diff
@@ -171,15 +171,6 @@ ReductionParams innerPersistentHeuristic(
   const int64_t max_multi_reduction_factor = scheduler_utils::safeDiv(
       scheduler_utils::register_file_size, max_persistent_buffer_size);
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/scheduler/normalization.cpp:173]: Removing inner persistent changes for now.

File: torch/csrc/jit/codegen/cuda/scheduler/normalization.cpp [hunk 5]
```diff
@@ -483,45 +432,25 @@ ReductionParams innerPersistentHeuristic(
   }
 
   if (godim > 1) {
-    if (rparams.cross_grid_inner_reduction) {
-      rparams.grid_dim_iter_dom = ParallelType::BIDy;
+    rparams.grid_dim_iter_dom = ParallelType::BIDx;
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/scheduler/normalization.cpp:435]: Should have kept this change instead of reverting. Can we isolate this into a stand alone PR as it seems to be a fix.

File: torch/csrc/jit/codegen/cuda/scheduler/normalization.cpp [hunk 6]
```diff
@@ -593,7 +524,7 @@ ReductionParams outerPersistentHeuristic(
       (int64_t)16 / (int64_t)max_input_dtype_size,
       // Reduce unrolling if we have many inputs, start reduction at 4 inputs
       scheduler_utils::lastPow2(
-          scheduler_utils::safeDiv((int64_t)n_tensor_inputs, 2)));
+          scheduler_utils::safeDiv((int64_t)n_tensor_inputs, 4)));
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/scheduler/normalization.cpp:527]: `<<2` is the equivalent of dividing by `4` not `2`.

File: torch/csrc/jit/codegen/cuda/scheduler/normalization.cpp [hunk 7]
```diff
@@ -711,22 +645,25 @@ ReductionParams outerPersistentHeuristic(
...
 
-  // Increase bdimy until we hit the SM partition granularity
-  auto num_warp_per_partition =
-      ceilDiv(bdimx * bdimy, 32 * scheduler_utils::getGpuSmPartitions());
-  while (ceilDiv(
-             bdimx * (bdimy + 1), 32 * scheduler_utils::getGpuSmPartitions()) ==
-         num_warp_per_partition) {
-    bdimy++;
-  }
+  //   // Increase bdimy until we hit the SM partition granularity
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/scheduler/normalization.cpp:653]: I removed this, only because I didn't understand what you're trying to do here. On second thought, I assume you're just trying to avoid going across warp boundaries that would allocate more registers, but uncertain why.

Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/scheduler/normalization.cpp:653]: I agree we probably want to consider these boundaries, but we're already limiting the register file quite significantly (more so than this could inflate the registers by).

File: torch/csrc/jit/codegen/cuda/scheduler/normalization.cpp [hunk 8]
```diff
@@ -793,24 +734,30 @@ ReductionParams outerPersistentHeuristic(
...
-      iter_unroll_factor /= 2;
-    } else {
-      bdimx = ceilDiv(bdimx, 2);
-    }
-  }
-
+  std::cout << "7: " << bdimx << ", " << bdimy << std::endl;
+  // Disable for now, it can go from taking half a warp, to reducing it to a
+  // quarter warp. On cases where we're dealing with fp16 values and we're not
+  // unrolled, this would mean loads are only 8xfp16 words strided across tidy.
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/scheduler/normalization.cpp:746]: This ^^

File: torch/csrc/jit/codegen/cuda/scheduler/normalization.cpp [hunk 9]
```diff
@@ -819,28 +766,33 @@ ReductionParams outerPersistentHeuristic(
 
   ReductionParams rparams;
 
+  // Set gdimy.
+  if (!block_persist) {
+    gdimy = sm_required_per_norm_set;
+  }
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/scheduler/normalization.cpp:778]: Need to set gdimy before trying to figure out the size of gdimx

**Comment by naoyam** [comment]
Thanks! I saw performance regressions when disabled welfrod translations. It seemed it was actually better not to translate for outer-reduction cases, but for inner-reduction batch norms, I saw up to 20% regressions. I'll report it back after doublechecking what I saw.

File: torch/csrc/jit/codegen/cuda/scheduler/registry.cpp [hunk 10]
```diff
@@ -1243,30 +1237,39 @@ class PersistentKernelScheduler : public SchedulerEntry {
     const int64_t required_sm_per_norm =
         ceilDiv(persistent_buffer_size, scheduler_utils::register_file_size);
 
+    // If the persistence requires over half the device don't do grid
+    // persistence as we can't overlap the grid comms.
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/scheduler/registry.cpp:1241]: What is overlapped with the grid comms?

File: torch/csrc/jit/codegen/cuda/scheduler/normalization.cpp (hunk 5 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/scheduler/normalization.cpp:435]: As the grid persistence is revered, doesn't this change make any difference? Am I missing something?

File: torch/csrc/jit/codegen/cuda/scheduler/normalization.cpp (hunk 6 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/scheduler/normalization.cpp:527]: 😳

File: torch/csrc/jit/codegen/cuda/scheduler/normalization.cpp (hunk 7 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/scheduler/normalization.cpp:653]: This was just copied over from your original PR. I thought that it was meant to increase the number of threads within the limit of the `32 * #partitions` granularity as doing so would have no actual impact on resource usage.

File: torch/csrc/jit/codegen/cuda/scheduler/normalization.cpp (hunk 9 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/scheduler/normalization.cpp:778]: Yeah, thanks for the fix.

**Review by naoyam — APPROVED** [review]
Thanks for the PR.

**Comment by naoyam** [comment]
> I'm getting 142GB/s on this size right now, which is underwhelming, but is better than where it started at.

Do you know what is the bandwidth with the non-persistent scheduler?

File: torch/csrc/jit/codegen/cuda/scheduler/registry.cpp (hunk 10 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/scheduler/registry.cpp:1241]: Just different groups of blocks running at the same time. I just don't think we want all SM's participating in the same grid persistent allreduce

File: torch/csrc/jit/codegen/cuda/scheduler/normalization.cpp (hunk 5 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/scheduler/normalization.cpp:435]: I was trying to just remove the changes unrelated to outer persistent reductions since that's what we really need quickly.

File: torch/csrc/jit/codegen/cuda/scheduler/normalization.cpp (hunk 7 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/scheduler/normalization.cpp:653]: That's probably right, we can revisit after we get a bit more out of the current available perf.

File: torch/csrc/jit/codegen/cuda/scheduler/normalization.cpp [hunk 11]
```diff
@@ -711,22 +645,25 @@ ReductionParams outerPersistentHeuristic(
...
+  //              scheduler_utils::getGpuSmPartitions()) ==
+  //          num_warp_per_partition) {
+  //     bdimy++;
+  //   }
 
   bool vectorize = false;
 
+  std::cout << "4: " << bdimx << ", " << bdimy << std::endl;
   // Move unrolling factor into vectorization upto vectorization limit.
   if (vectorize_factor > 1 && iter_unroll_factor > 1) {
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/scheduler/normalization.cpp:667]: This block needs to be protected because it's setting `iter_unroll_factor` but not checking `max_multi_reduction_factor`. We need to change this block to:
```
  if (vectorize_factor > 1 && iter_unroll_factor > 1) {
    vectorize = true;
    iter_unroll_factor = std::min(
        scheduler_utils::lastPow2(iter_unroll_factor),
        (int64_t)vectorize_factor);
    iter_unroll_factor = std::min(
        iter_unroll_factor,
        scheduler_utils::safeDiv(max_multi_reduction_factor, bdimx));
  }
  ```

**Comment by csarofeen** [comment]
For some reason I seem to not have a response block to your inline comments:

> As the grid persistence is revered, doesn't this change make any difference? Am I missing something?

Just ignore any changes I made to inner persistent for now. I was just trying to isolate out cleanup from heuristic changes. Just focus on the outer persistent stuff, and you can push fixes to inner persistent in another PR.

> This was just copied over from your original PR. I thought that it was meant to increase the number of threads within the limit of the 32 * #partitions granularity as doing so would have no actual impact on resource usage.

Yeah I think you're right, let's revisit this later to see if it's actually valuable.



>Do you know what is the bandwidth with the non-persistent  scheduler?

Let me pull it.

**Comment by csarofeen** [comment]
With grid persistence:
kernel1 run in 4.37616 ms, achieved: 122.7 GB/s

Without:
kernel1 run in 1.08733 ms, achieved: 246.952 GB/s
kernel2 run in 0.892608 ms, achieved: 601.5 GB/s

**Comment by csarofeen** [comment]
there's definitely some register usage issues. I think we need to check on `nvfuser_zero` insertion, as manually adding more of those and zero update was reducing register pressure.
